File: P32 Longest Valid Parentheses/Solution.java
class Solution {
    // Time : O(n), n is the length of the string.
    // Space: O(n), the size of the stack can be up to n. It could be improved to O(1) by scaning the string from left to right and right to left
    public int longestValidParentheses(String s) {
        int result = 0;
        Stack<Integer> stack = new Stack<>();
        int pointer = 0;
        stack.push(-1);
        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '(') {
                stack.push(i);
            } else {
                stack.pop();
                if(stack.empty()) {
                   stack.push(i); 
                } else {
                    result = Math.max(result, i - stack.peek());
                }
            }
        }
        return result;
    }
}

File: P32 Longest Valid Parentheses/solution.py
class Solution:
    # Time : O(n), n is the length of the string.
    # Space: O(n), the size of the stack can be up to n. It could be improved to O(1) by scaning the string from left to right and right to left
    def longestValidParentheses(self, s: str) -> int:
        result = 0
        stack = [-1]
        for i in range(len(s)):
            if s[i] == '(':
                stack.append(i)
            else:
                stack.pop()
                if len(stack) == 0:
                    stack.append(i)
                else:
                    result = max(result, i - stack[-1])
        return result
